### DISCUSSION TITLE
easier immutable record fields

### DISCUSSION THREAD

**Discussion by Eschryn** [dotnet/csharplang#5225] [Language Ideas] [answered]
Currently when something is written in the parameters of the record it is mutable with with. When you try to create a get only Property it is not settable (via default constructor).

Idea: have the ability to somehow include a get only properties in the constructor of a record - one Idea how this could be done:
```cs
record Person(
    readonly int Id,
    string Name
);
```

**Comment by spydacarnage** [comment]
`with` doesn't mutate - it creates a new object with the new values.

**Reply by svick** [reply]
I think that's just a matter of terminology. [What `with` does is sometimes called "non-destructive mutation",](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/records?WT.mc_id=DOP-MVP-5002735#characteristics-of-records) and this request is about limiting that: you want people to be able to write `person with { Name = newName }`, but you want to prohibit `person with { Id = newId }`. (Though you can't prevent them from writing `new Person(newId, person.Name)`, so it's not airtight.)

**Comment by DavidArno** [comment, marked as answer]
@Eschryn,

What would be the purpose with preventing `Id` in your example from being mutated via a wither? As @svick points out, even if you could define a record like that, there's nothing to stop someone just creating a new `Person` instance:
```cs
var p1 = new Person(1, "Tom");
var p2 = new Person(2, p1,Name);
```

But if you really want to do this, one way around it is to avoid a primary constructor and to use a factory method and properties. That way, you can mark any properties that you don't want the wither accessing as `private init`.  
```cs
rrecord Person
{
    private Person() {}

    public int Id { get; private init; }

    public string Name { get; init; }
    
    public static Person Create(int id, string name) => new Person { Id = id, Name = name };
}

var p1 = Person.Create(1, "Liz");
var p2 = p1 with { Id = 2 }; // 'Person.Id' cannot be used in this context because the set accessor is inaccessible
var p3 = p1 with { Name = "Ada" }; // All fine
```
Still not sure what this really gain you though as I can still call `Person.Create` as many times as I want.

**Reply by Eschryn** [reply]
It was more a terminology thing, though I had a use case. New as in creating entirely new and with does "non-destructive mutation". If I create a new Object it is a new Object anyway. With implies more that it's a copy of the related object with different values. That was my original thought.

**Reply by jnm2** [reply]
My reason for blocking `with`-ability for individual properties is to make it harder to forget to update two or three properties which should be updated in conjunction. This reason wouldn't apply to `new`.

**Comment by jnm2** [comment]
Here's how you can do it in C# 9 syntax:

```cs
record Person(
    int Id,
    string Name)
{
    public int Id { get; } = Id;
}
```

If you want `Person` to be able to update the `Id` property itself using `with`, then you can do this:


```cs
record Person(
    int Id,
    string Name)
{
    public int Id { get; private init; } = Id;
}
```

**Reply by DavidArno** [reply]
Nice. That's way simpler than my code above. 👍
